simulator/src/scrypto_bindgen/translation.rs

```diff
@@ -0,0 +1,310 @@
+            ref_types: RefTypes::NORMAL | RefTypes::DIRECT_ACCESS,
+            receiver: Receiver::SelfRefMut,
+        }) => ast::FnType::Method {
+            is_mutable_receiver: true,
+        },
+        None => ast::FnType::Function,
+        _ => panic!("Invalid BitFlags for RefTypes"),
+    };
+    let ident = ident!(&schema_stub.ident);
+
+    let inputs = schema_stub
+        .arguments
+        .into_iter()
+        .map(|(arg_name, arg_type_index)| {
+            type_name(&arg_type_index, blueprint_name.as_str(), schema_resolver).map(|type_name| {
+                (
+                    ident!(&arg_name),
+                    syn::parse2(type_name.parse().unwrap()).unwrap(),
+                )
+            })
+        })
+        .collect::<Result<_, _>>()?;
+    let output = syn::parse2(
+        type_name(
+            &schema_stub.returns,
+            blueprint_name.as_str(),
+            schema_resolver,
+        )?
+        .parse()
+        .unwrap(),
+    )
+    .unwrap();
+
+    Ok(ast::InvocationFn {
+        inputs,
+        fn_type,
+        ident,
+        output,
+    })
+}
+
+fn type_name<S>(
+    type_identifier: &ScopedTypeId,
+    blueprint_name: &str,
+    schema_resolver: &S,
+) -> Result<String, schema::SchemaError>
+where
+    S: schema::PackageSchemaResolver,
+{
+    let type_kind = schema_resolver.resolve_type_kind(type_identifier)?;
+    let type_metadata = schema_resolver.resolve_type_metadata(type_identifier)?;
+    let type_validation = schema_resolver.resolve_type_validation(type_identifier)?;
+    let type_ident = type_metadata.get_name_string();
+
+    let name = match type_kind {
+        TypeKind::Any => type_ident.unwrap_or("ScryptoValue".to_owned()),
+        TypeKind::Bool => type_ident.unwrap_or("bool".to_owned()),
+        TypeKind::I8 => type_ident.unwrap_or("i8".to_owned()),
+        TypeKind::I16 => type_ident.unwrap_or("i16".to_owned()),
+        TypeKind::I32 => type_ident.unwrap_or("i32".to_owned()),
+        TypeKind::I64 => type_ident.unwrap_or("i64".to_owned()),
+        TypeKind::I128 => type_ident.unwrap_or("i128".to_owned()),
+        TypeKind::U8 => type_ident.unwrap_or("u8".to_owned()),
+        TypeKind::U16 => type_ident.unwrap_or("u16".to_owned()),
+        TypeKind::U32 => type_ident.unwrap_or("u32".to_owned()),
+        TypeKind::U64 => type_ident.unwrap_or("u64".to_owned()),
+        TypeKind::U128 => type_ident.unwrap_or("u128".to_owned()),
+        TypeKind::String => type_ident.unwrap_or("String".to_owned()),
+        TypeKind::Array { element_type } => type_ident.unwrap_or(format!(
+            "Vec<{}>",
+            type_name(
+                &ScopedTypeId(type_identifier.0, element_type),
+                blueprint_name,
+                schema_resolver
+            )?
+        )),
+        TypeKind::Tuple { field_types } => type_ident.unwrap_or(if field_types.is_empty() {
+            "()".to_owned()
+        } else {
+            format!(
+                "({},)",
```

Comment:
The extra trailing `,` is to deal with single element tuple?

Reply:
Yup exactly. To the compiler a `(T)` is just `T`. But `(T,)` is a single element tuple. So, we force them to be tuples with the extra trailing comma.